### ISSUE TITLE
[Introspection] Table not found due to missing '_'

### ISSUE THREAD

**Issue by williamluke4** [prisma/prisma2#1289]
Could be related to #1057 and #1067

### Steps to Reproduce
1.  Use the prisma1 schema found here [prisma/prisma2-private#15](https://github.com/prisma/prisma2-private/issues/15)
2. Deploy it to a postgres db
3. Run `prisma2 introspect`

### Thoughts
In the log, it says `Table ChatRoomFromTrip not found` and indeed `ChatRoomFromTrip` is not in the database but `_ChatRoomFromTrip` is present. I would assume that the `_` has been removed at some point during the introspection?

### Log
```
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Found no corresponding foreign key    
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Handling column Column { name: "A", tpe: ColumnType { raw: "varchar", family: String, arity: Required }, default: None, auto_increment: false }    
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Calculating field type for 'A'    
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Found corresponding foreign key    
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Handling column Column { name: "B", tpe: ColumnType { raw: "varchar", family: String, arity: Required }, default: None, auto_increment: false }    
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Calculating field type for 'B'    
Jan 06 18:59:45.426 DEBUG Introspect{url=UrlInput { url: "postgresql://prisma:prisma@localhost:5434/prisma?schema=API%24production" }}: sql_introspection_connector::calculate_datamodel: Found corresponding foreign key    

{"message":"[src/libcore/result.rs:1165:5] called `Result::unwrap()` on an `Err` value: \"Table ChatRoomFromTrip not found\"","backtrace":"stack backtrace:\n   0: user_facing_errors::UnknownError::new_in_panic_hook\n   1: user_facing_errors::panic_hook::set_panic_hook::{{closure}}\n   2: std::panicking::rust_panic_with_hook\n             at src/libstd/panicking.rs:468\n   3: std::panicking::continue_panic_fmt\n             at src/libstd/panicking.rs:373\n   4: rust_begin_unwind\n             at src/libstd/panicking.rs:302\n   5: core::panicking::panic_fmt\n             at src/libcore/panicking.rs:139\n   6: core::result::unwrap_failed\n             at src/libcore/result.rs:1165\n   7: sql_schema_describer::SqlSchema::table_bang\n   8: sql_introspection_connector::calculate_datamodel::calculate_model\n   9: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  10: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  11: tokio::runtime::enter::Enter::block_on\n  12: std::thread::local::LocalKey<T>::with\n  13: std::thread::local::LocalKey<T>::with\n  14: tokio::runtime::spawner::Spawner::enter\n  15: tokio::runtime::time::variant::with_default\n  16: std::thread::local::LocalKey<T>::with\n  17: <introspection_engine::rpc::RpcImpl as introspection_engine::rpc::rpc_impl_Rpc::gen_server::Rpc>::introspect\n  18: <jsonrpc_core::delegates::DelegateAsyncMethod<T,F> as jsonrpc_core::calls::RpcMethod<M>>::call\n  19: <futures::future::lazy::Lazy<F,R> as futures::future::Future>::poll\n  20: futures::future::chain::Chain<A,B,C>::poll\n  21: <futures::future::then::Then<A,B,F> as futures::future::Future>::poll\n  22: <futures::future::map::Map<A,F> as futures::future::Future>::poll\n  23: <futures::future::either::Either<A,B> as futures::future::Future>::poll\n  24: <futures::future::map::Map<A,F> as futures::future::Future>::poll\n  25: <futures::future::map_err::MapErr<A,F> as futures::future::Future>::poll\n  26: <futures::stream::and_then::AndThen<S,F,U> as futures::stream::Stream>::poll\n  27: <futures::stream::forward::Forward<T,U> as futures::future::Future>::poll\n  28: <futures::future::map::Map<A,F> as futures::future::Future>::poll\n  29: <futures::future::map_err::MapErr<A,F> as futures::future::Future>::poll\n  30: futures::task_impl::Spawn<T>::poll_future_notify\n  31: std::panicking::try::do_call\n  32: __rust_maybe_catch_panic\n             at src/libpanic_unwind/lib.rs:78\n  33: tokio_threadpool::task::Task::run\n  34: tokio_threadpool::worker::Worker::run_task\n  35: tokio_threadpool::worker::Worker::run\n  36: tokio_reactor::with_default\n  37: tokio::runtime::threadpool::builder::Builder::build::{{closure}}\n  38: std::thread::local::LocalKey<T>::with\n  39: std::sys_common::backtrace::__rust_begin_short_backtrace\n  40: std::panicking::try::do_call\n  41: __rust_maybe_catch_panic\n             at src/libpanic_unwind/lib.rs:78\n  42: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  43: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once\n             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:942\n  44: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once\n             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:942\n      std::sys_common::thread::start_thread\n             at src/libstd/sys_common/thread.rs:13\n      std::sys::unix::thread::Thread::new::thread_start\n             at src/libstd/sys/unix/thread.rs:79\n  45: start_thread\n  46: __clone\n"}
```

**Comment by pantharshit00** [comment]
I can confirm this bug in alpha ie: `prisma2@2.0.0-alpha.501, binary version: c6c401330e769e9813b07d3a6815001ca3daa9e1`

```
Error: Error in introspection engine: Jan 08 19:56:40.309  INFO Introspect{url=UrlInput { url: "postgresql://harshit:Harshit00@localhost:5432/issue_1289?schema=API$production" }}: quaint::single: Starting a postgresql pool with 1 connections.
{"is_panic":true,"message":"[src/libcore/result.rs:1165:5] called `Result::unwrap()` on an `Err` value: \"Table ChatRoomFromTrip not found\"","backtrace":"stack backtrace:\n   0: backtrace::backtrace::trace\n   1: backtrace::capture::Backtrace::new\n   2: user_facing_errors::Error::new_in_panic_hook\n   3: user_facing_errors::panic_hook::set_panic_hook::{{closure}}\n   4: std::panicking::rust_panic_with_hook\n   5: std::panicking::continue_panic_fmt\n   6: rust_begin_unwind\n   7: core::panicking::panic_fmt\n   8: core::result::unwrap_failed\n   9: sql_schema_describer::SqlSchema::table_bang\n  10: sql_introspection_connector::calculate_datamodel::calculate_model\n  11: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  12: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  13: tokio::runtime::enter::Enter::block_on\n  14: std::thread::local::LocalKey<T>::with\n  15: std::thread::local::LocalKey<T>::with\n  16: tokio::runtime::spawner::Spawner::enter\n  17: tokio::runtime::time::variant::with_default\n  18: std::thread::local::LocalKey<T>::with\n  19: <introspection_engine::rpc::RpcImpl as introspection_engine::rpc::rpc_impl_Rpc::gen_server::Rpc>::introspect\n  20: <jsonrpc_core::delegates::DelegateAsyncMethod<T,F> as jsonrpc_core::calls::RpcMethod<M>>::call\n  21: <futures::future::lazy::Lazy<F,R> as futures::future::Future>::poll\n  22: futures::future::chain::Chain<A,B,C>::poll\n  23: <futures::future::then::Then<A,B,F> as futures::future::Future>::poll\n  24: <futures::future::map::Map<A,F> as futures::future::Future>::poll\n  25: <futures::future::either::Either<A,B> as futures::future::Future>::poll\n  26: <futures::future::map::Map<A,F> as futures::future::Future>::poll\n  27: <futures::future::map_err::MapErr<A,F> as futures::future::Future>::poll\n  28: <futures::stream::and_then::AndThen<S,F,U> as futures::stream::Stream>::poll\n  29: <futures::stream::forward::Forward<T,U> as futures::future::Future>::poll\n  30: <futures::future::map::Map<A,F> as futures::future::Future>::poll\n  31: <futures::future::map_err::MapErr<A,F> as futures::future::Future>::poll\n  32: futures::task_impl::Spawn<T>::poll_future_notify\n  33: std::panicking::try::do_call\n  34: __rust_maybe_catch_panic\n  35: tokio_threadpool::task::Task::run\n  36: tokio_threadpool::worker::Worker::run_task\n  37: tokio_threadpool::worker::Worker::run\n  38: tokio_reactor::with_default\n  39: tokio::runtime::threadpool::builder::Builder::build::{{closure}}\n  40: std::thread::local::LocalKey<T>::with\n  41: std::sys_common::backtrace::__rust_begin_short_backtrace\n  42: std::panicking::try::do_call\n  43: __rust_maybe_catch_panic\n  44: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  45: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once\n  46: std::sys::unix::thread::Thread::new::thread_start\n  47: _pthread_start\n"}
```

**Comment by janpio** [comment]
Next steps: Reproduce again and see what happens.

**Comment by williamluke4** [comment]
Just tested using  `prisma2@2.0.0-alpha.710, binary version: d8a1685e3537e330f73cd874b9470f3a21815dd7` and this particular issue seems to be fixed. So I'm going to close it :)

**Comment by janpio** [comment]
I just introspected the database schema from https://github.com/prisma/prisma2-private/issues/15 again with `prisma2 introspect` of `prisma2@2.0.0-preview021, binary version: 4f8eb5bd9628db47402dc855ad0b6c7cfc915d46` and could not observe this problem any more.

The model mentioned above now looks like this:
```
model ChatRoomFromTrip {
  A ChatRoom @relation("ChatRoomTo_ChatRoomFromTrip")
  B Trip     @relation("TripTo_ChatRoomFromTrip")

  @@map("_ChatRoomFromTrip")
}
```

The `@map` is necessary because we do not support models starting with underscore. The relation in `ChatRoom` looks like this for example:

```
  ChatRoomFromTrips ChatRoomFromTrip[] @map("_ChatRoomFromTrips") @relation("ChatRoomTo_ChatRoomFromTrip", references: [A])
```

---

But I got a very different error now:
```
Introspection failed: Introspected schema can't be parsed.
Error: Schema parsing
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "No id field defined! Model: FlightRouteDestinationAirportsOnFlightCharter"', src\libcore\result.rs:1165:5
```

You already know _that_ one from #1447 and we are tracking this in issues like https://github.com/prisma/prisma2/issues/1113 (and others, because there are multiple variants of this) - so I think we can safely close this issue and continue in the other ones.

Thanks again so much for providing us your schema - that helped super much not to first understand the problem in the first place, but also confirm now that the problem is actually fixed <3
